docs(notebooks): add partial folding guide

ritajitmajumdar1 opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [x ] I have added the tests to cover my changes.
- [x ] I have updated the documentation accordingly.
- [x ] I have read the CONTRIBUTING document.
-->

### Summary

This pull request is for the how-to guide on partial folding with ZNE which now works with Qiskit 1.0.



### Details and comments

Requesting feedback

pedrorrivero commented:
Hi @ritajitmajumdar1 ,

Please add the `wire-cutting` guide in a separate branch and open a new PR.

miamico reviewed (changes requested):
The guide looks great! Some minor comments:

- This function `build_compute_uncompute_qaoa_circuit` has a PR to be hosted in the repo, let's keep it like this for now but remember to import it once everything is merged.

- The noise factors used in the regular zne comparison `zne_options.resilience.noise_factors = [1, 3, 5]` are too large (expectation values are 3 are already 0), can we try [1,2,3] instead?

ritajitmajumdar1 commented:
@ashsaki I have just updated the notebook as per your reviews. @miamico your reviews on using noise factors [1,2,3] are also incorporated. Although, I may be wrong, isn't [1,3,5] the default values in Primitives?

ritajitmajumdar1 commented:
@ashsaki @pedrorrivero @miamico, I was testing out the ZNE multiple times. Here is the code snippet in case I am doing something wrong:

```
estimator = Estimator(session=session)
estimator.options.resilience_level=2 # applies ZNE
estimator.options.resilience.zne.noise_factors=[1,1.2,1.4]
estimator.options.resilience.zne.extrapolator='linear'
estimator.options.default_shots=5000

job = estimator.run([(trans_qaoa, trans_obs)])
```

I am attaching the figure after running this code with the same trans_qaoa and trans_obs for 20 trials, and see the variance in the extrapolated result. Is such a variance acceptable?
![test](https://github.com/IBM-Quantum-Technical-Enablement/quantum-enablement/assets/122992142/5b2423c5-9a12-4647-bec3-881c07f76fda)

miamico commented:
I'm not surprised to see somewhat high variance in the extrapolated values though. This will depend on the fit on the measured expectation values at the various noise factor and if those are a bit noisy, the fit will end up being noisy as well. If it was possible to increment the number of circuits used to get the approximate partial increase in noise factor, I'd expect to see more stable extrapolations.

ashsaki commented:
@ritajitmajumdar1 let me know if your done testing on the HW and make sure to check if the returned `zne_noise_factors` is `0.0`. Once you confirm these, I will approve the PR.

ashsaki commented on docs/notebooks/partial_folding.ipynb:
```diff
@@ -0,0 +1,815 @@
+   "outputs": [
+    {
+     "data": {
+      "image/png": "iVBORw0KGgoAAAANSUhEUgAAGFkAAAf5CAYAAACdhYJBAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/H5lhTAAAACXBIWXMAAA9hAAAPYQGoP6dpAAEAAElEQVR4nOzdeZyN9f//8eeZBsNYxtjN2LJmLBMleyak7ERkCaURlSSftFkqFaWQ7LILfe0fWxuKkhbJTvqYxjCTLIOZYSxzfn/4mUxmzDnHua6zXI/77eaWmXO9z3menOv1us513ud62+x2u10AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABeLsDTAQAAAAAAAAA ...
```
Simplify to
```suggestion
    "trans_obs = observable.apply_layout(trans_qaoa.layout)"
```

pedrorrivero reviewed (changes requested):
Thanks so much @ritajitmajumdar1 ! Great work overall

A few comments to address:

- [x] _Noise amplification_ (i.e. not ZNE) is categorized in the following types:
  - Analogue → Hardware based (e.g. pulse)
  - Digital → Circuit based
    - Foldings
    - PEA
- [x] Add links to references
- [x] Adjust your notation to match that in the scaling notebook(s) (e.g. $U(U^\dagger U)^f$, $\lambda = 2f + 1$)
- [x] Make sure that you include all contents from the scaling notebook(s) here.
- [x] Update the circuit building functions to latest
- [x] Why $p=5$ but two-qubit depth is 40? It should be 20 after the compute-uncompute. I believe this is related to transpilation of parametrized circuits.
- [x] I think it would be beneficial to expand a bit on the theory of partial foldings, maybe point to the dZNE best practices paper
- [x] Deactivate measurement and gate twirling, as well as TREX from resilience level 2
- [ ] Add a section title for the references

ritajitmajumdar1 commented:
Made the changes. I have added links to related paper within the text, so removed reference section.

pedrorrivero reviewed (approved):
Thanks @ritajitmajumdar1, good job! LGTM 😊
